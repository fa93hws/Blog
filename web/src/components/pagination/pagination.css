.container {
  display: flex;
}

.currentNumber {
  color: red;
}
