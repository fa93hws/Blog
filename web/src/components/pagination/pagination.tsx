import * as React from 'react';
import * as classnames from 'classnames';
import styles from './pagination.css';

const STEP_SIZE = 3;

type PaginationProps = {
  totalPage: number;
  currentPage: number;
  getLink(page: number): string;
};

function verifyProps(props: PaginationProps) {
  if (props.currentPage < 0) {
    throw new Error('currentPage must be greater than 0');
  }
  if (props.currentPage > props.totalPage) {
    throw new Error('currentPage must not be greater than totalPage');
  }
}

type NumberProps = number | 'hellip';

function generateItems({
  current,
  total,
}: {
  current: number;
  total: number;
}): NumberProps[] {
  if (total <= 7) {
    return Array.from(new Array(total), (_, idx) => idx + 1);
  }
  if (current <= 1 + STEP_SIZE) {
    return [1, 2, 3, 4, 5, 'hellip', total];
  }
  if (current >= total - STEP_SIZE) {
    return [1, 'hellip', total - 4, total - 3, total - 2, total - 1, total];
  }
  return [1, 'hellip', current - 1, current, current + 1, 'hellip', total];
}

const NumberButton = ({
  number,
  isCurrent,
}: {
  number: number;
  isCurrent: boolean;
}) => (
  <div
    className={classnames({
      [styles.currentNumber]: isCurrent,
    })}
  >
    {number}
  </div>
);
const Hellip = () => <span>&hellip;</span>;
const Item = React.memo(
  ({ current, item }: { current: number; item: number | 'hellip' }) => {
    if (item === 'hellip') {
      return <Hellip />;
    }
    return <NumberButton isCurrent={current === item} number={item} />;
  },
);

export const Pagination = (props: PaginationProps) => {
  verifyProps(props);
  const items = generateItems({
    current: props.currentPage,
    total: props.totalPage,
  });
  return (
    <ul className={styles.container}>
      {items.map((item, idx) => (
        <li key={idx}>
          <Item item={item} current={props.currentPage} />
        </li>
      ))}
    </ul>
  );
};
