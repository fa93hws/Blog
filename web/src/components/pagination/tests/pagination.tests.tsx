import * as React from 'react';
import { Pagination } from '../pagination';

describe('Pagination', () => {
  it.each([1, 4, 5, 6, 7, 10])('renders %d/10', (idx: number) => {
    expect(
      <Pagination currentPage={idx} totalPage={10} getLink={jest.fn()} />,
    ).toMatchRenderedSnapshot();
  });

  it.each([1, 4, 7])('renders %d/7', (idx: number) => {
    expect(
      <Pagination currentPage={idx} totalPage={7} getLink={jest.fn()} />,
    ).toMatchRenderedSnapshot();
  });
});
