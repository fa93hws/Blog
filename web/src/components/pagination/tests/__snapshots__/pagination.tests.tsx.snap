// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders 1/7 1`] = `
<ul
  class="container"
>
  <li>
    <div
      class="currentNumber"
    >
      1
    </div>
  </li>
  <li>
    <div
      class=""
    >
      2
    </div>
  </li>
  <li>
    <div
      class=""
    >
      3
    </div>
  </li>
  <li>
    <div
      class=""
    >
      4
    </div>
  </li>
  <li>
    <div
      class=""
    >
      5
    </div>
  </li>
  <li>
    <div
      class=""
    >
      6
    </div>
  </li>
  <li>
    <div
      class=""
    >
      7
    </div>
  </li>
</ul>
`;

exports[`Pagination renders 1/10 1`] = `
<ul
  class="container"
>
  <li>
    <div
      class="currentNumber"
    >
      1
    </div>
  </li>
  <li>
    <div
      class=""
    >
      2
    </div>
  </li>
  <li>
    <div
      class=""
    >
      3
    </div>
  </li>
  <li>
    <div
      class=""
    >
      4
    </div>
  </li>
  <li>
    <div
      class=""
    >
      5
    </div>
  </li>
  <li>
    <span>
      …
    </span>
  </li>
  <li>
    <div
      class=""
    >
      10
    </div>
  </li>
</ul>
`;

exports[`Pagination renders 4/7 1`] = `
<ul
  class="container"
>
  <li>
    <div
      class=""
    >
      1
    </div>
  </li>
  <li>
    <div
      class=""
    >
      2
    </div>
  </li>
  <li>
    <div
      class=""
    >
      3
    </div>
  </li>
  <li>
    <div
      class="currentNumber"
    >
      4
    </div>
  </li>
  <li>
    <div
      class=""
    >
      5
    </div>
  </li>
  <li>
    <div
      class=""
    >
      6
    </div>
  </li>
  <li>
    <div
      class=""
    >
      7
    </div>
  </li>
</ul>
`;

exports[`Pagination renders 4/10 1`] = `
<ul
  class="container"
>
  <li>
    <div
      class=""
    >
      1
    </div>
  </li>
  <li>
    <div
      class=""
    >
      2
    </div>
  </li>
  <li>
    <div
      class=""
    >
      3
    </div>
  </li>
  <li>
    <div
      class="currentNumber"
    >
      4
    </div>
  </li>
  <li>
    <div
      class=""
    >
      5
    </div>
  </li>
  <li>
    <span>
      …
    </span>
  </li>
  <li>
    <div
      class=""
    >
      10
    </div>
  </li>
</ul>
`;

exports[`Pagination renders 5/10 1`] = `
<ul
  class="container"
>
  <li>
    <div
      class=""
    >
      1
    </div>
  </li>
  <li>
    <span>
      …
    </span>
  </li>
  <li>
    <div
      class=""
    >
      4
    </div>
  </li>
  <li>
    <div
      class="currentNumber"
    >
      5
    </div>
  </li>
  <li>
    <div
      class=""
    >
      6
    </div>
  </li>
  <li>
    <span>
      …
    </span>
  </li>
  <li>
    <div
      class=""
    >
      10
    </div>
  </li>
</ul>
`;

exports[`Pagination renders 6/10 1`] = `
<ul
  class="container"
>
  <li>
    <div
      class=""
    >
      1
    </div>
  </li>
  <li>
    <span>
      …
    </span>
  </li>
  <li>
    <div
      class=""
    >
      5
    </div>
  </li>
  <li>
    <div
      class="currentNumber"
    >
      6
    </div>
  </li>
  <li>
    <div
      class=""
    >
      7
    </div>
  </li>
  <li>
    <span>
      …
    </span>
  </li>
  <li>
    <div
      class=""
    >
      10
    </div>
  </li>
</ul>
`;

exports[`Pagination renders 7/7 1`] = `
<ul
  class="container"
>
  <li>
    <div
      class=""
    >
      1
    </div>
  </li>
  <li>
    <div
      class=""
    >
      2
    </div>
  </li>
  <li>
    <div
      class=""
    >
      3
    </div>
  </li>
  <li>
    <div
      class=""
    >
      4
    </div>
  </li>
  <li>
    <div
      class=""
    >
      5
    </div>
  </li>
  <li>
    <div
      class=""
    >
      6
    </div>
  </li>
  <li>
    <div
      class="currentNumber"
    >
      7
    </div>
  </li>
</ul>
`;

exports[`Pagination renders 7/10 1`] = `
<ul
  class="container"
>
  <li>
    <div
      class=""
    >
      1
    </div>
  </li>
  <li>
    <span>
      …
    </span>
  </li>
  <li>
    <div
      class=""
    >
      6
    </div>
  </li>
  <li>
    <div
      class="currentNumber"
    >
      7
    </div>
  </li>
  <li>
    <div
      class=""
    >
      8
    </div>
  </li>
  <li>
    <div
      class=""
    >
      9
    </div>
  </li>
  <li>
    <div
      class=""
    >
      10
    </div>
  </li>
</ul>
`;

exports[`Pagination renders 10/10 1`] = `
<ul
  class="container"
>
  <li>
    <div
      class=""
    >
      1
    </div>
  </li>
  <li>
    <span>
      …
    </span>
  </li>
  <li>
    <div
      class=""
    >
      6
    </div>
  </li>
  <li>
    <div
      class=""
    >
      7
    </div>
  </li>
  <li>
    <div
      class=""
    >
      8
    </div>
  </li>
  <li>
    <div
      class=""
    >
      9
    </div>
  </li>
  <li>
    <div
      class="currentNumber"
    >
      10
    </div>
  </li>
</ul>
`;
